src/block.rs

```diff
@@ -157,55 +159,103 @@ impl<G: Generator> Drop for BlockRng<G> {
     }
 }
 
-impl<W: Copy + Default, const N: usize, G: Generator<Output = [W; N]>> BlockRng<G> {
+impl<W: Word + Default, const N: usize, G: Generator<Output = [W; N]>> BlockRng<G> {
     /// Create a new `BlockRng` from an existing RNG implementing
     /// `Generator`. Results will be generated on first use.
     #[inline]
     pub fn new(core: G) -> BlockRng<G> {
-        BlockRng {
-            core,
-            index: N,
-            results: [W::default(); N],
+        let mut results = [W::default(); N];
+        results[0] = W::from_usize(N);
+        BlockRng { core, results }
+    }
+
+    /// Reconstruct from a core and a remaining-results buffer.
+    ///
+    /// This may be used to deserialize using a `core` and the output of
+    /// [`Self::remaining_results`].
+    ///
+    /// Returns `None` if `remaining_results` is too long.
+    pub fn reconstruct(core: G, remaining_results: &[W]) -> Option<Self> {
+        let mut results = [W::default(); N];
+        if remaining_results.len() < N {
+            let index = N - remaining_results.len();
+            results[index..].copy_from_slice(remaining_results);
+            results[0] = W::from_usize(index);
+            Some(BlockRng { results, core })
+        } else {
+            None
         }
     }
 }
 
-impl<W: Clone, const N: usize, G: Generator<Output = [W; N]>> BlockRng<G> {
+impl<W: Word, const N: usize, G: Generator<Output = [W; N]>> BlockRng<G> {
     /// Get the index into the result buffer.
     ///
     /// If this is equal to or larger than the size of the result buffer then
     /// the buffer is "empty" and `generate()` must be called to produce new
     /// results.
     #[inline(always)]
     pub fn index(&self) -> usize {
-        self.index
+        self.results[0].into_usize()
+    }
+
+    #[inline(always)]
+    fn set_index(&mut self, index: usize) {
+        debug_assert!(0 < index && index <= N);
+        self.results[0] = W::from_usize(index);
     }
 
     /// Reset the number of available results.
     /// This will force a new set of results to be generated on next use.
     #[inline]
     pub fn reset(&mut self) {
-        self.index = N;
+        self.set_index(N);
     }
 
-    /// Generate a new set of results immediately, setting the index to the
-    /// given value.
+    /// Updates the index and buffer contents
+    ///
+    /// If `index == 0`, this marks the buffer as "empty", causing generation on
+    /// next use.
+    ///
+    /// If `index > 0`, this generates a new block immediately then sets the
+    /// index.
     #[inline]
     pub fn generate_and_set(&mut self, index: usize) {
```

Comment:
This is a change of behaviour, but should only be observable if the caller uses `generate_and_set(0)` without immediately generating any output *and* cares that `Generator::generate` was not called. As such, I didn't bother mentioning this in the CHANGELOG.

I considered renaming it but decided not to.